data_and_models/pipelines/ner/clean.py

```diff
@@ -0,0 +1,121 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+"""
+This script cleans annotations exported with Prodigy as .jsonl files.
+
+Run 'analyze.py' before to identify relevant cleaning rules.
+
+Usage: clean.py [OPTIONS] INPUT_PATH KEEP_LABEL [RENAME_INTO]
+
+First, this script keeps only valid texts from the annotations (see next paragraph).
+Second, this script normalizes entity labels (use upper case for labels in lower case).
+Third, this script keeps only the given entity label and renames it if necessary.
+
+Valid texts are texts which have been accepted, have or not spans (i.e. entities), and
+have a unique 'input_hash'.
+
+The output is one file. The file is put in the same directory as the input file. It is
+named following the pattern 'annotations_<entity_label>.jsonl'.
+"""
+
+from collections import defaultdict
+from copy import deepcopy
+from pathlib import Path
+from typing import Any, Dict, Optional
+
+import srsly
+import typer
+
+
+def is_valid(example: Dict[str, Any], duplicated_hashes: set) -> bool:
+    accepted = example["answer"] == "accept"
+    unique_hash = example["_input_hash"] not in duplicated_hashes
+    return all((accepted, unique_hash))
+
+
+def main(
+    input_path: Path = typer.Argument(..., exists=True, dir_okay=False),
+    keep_label: str = typer.Argument(...),
+    rename_into: Optional[str] = typer.Argument(None),
+):
+    keep_label = keep_label.upper()
+    rename_into = None if rename_into is None else rename_into.upper()
+
+    print("Read annotations...")
+    corpus = list(srsly.read_jsonl(input_path))
+    print(f"...read {len(corpus)} texts")
+
+    print("Identify duplicated 'input_hash'...")
+    seen_hashes = set()
+    duplicated_hashes = set()
+    for x in corpus:
+        hash = x["_input_hash"]
+        if hash in seen_hashes:
+            duplicated_hashes.add(hash)
+        else:
+            seen_hashes.add(hash)
+    hashes_count = len(duplicated_hashes)
+    print(f"...identified {hashes_count} hash(es)")
+    if hashes_count > 0:
+        debug = " ".join(map(str, sorted(duplicated_hashes)))
+        print(f">>> {debug}")
+
+    print("Keep only valid texts...")
+    valid_texts = [x for x in corpus if is_valid(x, duplicated_hashes)]
+    print(f"...kept {len(valid_texts)} texts")
+    print(f"...that's {len(valid_texts) / len(corpus):.0%} of the total")
+
+    print("Normalize labels...")
+    total_spans = 0
+    normalized_spans = 0
+    for text in valid_texts:
+        for span in text.get("spans", []):
+            label = span["label"]
+            total_spans += 1
+            if label.islower():
```

Comment:
Maybe it's a bit more precise to write `if not label.isupper()`. Maybe we don't expect mixed-case labels at the moment, but I think the construction
```python
if not label.isupper():
    label = label.upper()
    ...
```
would exclude all potential trouble with mixed-case labels in the future.

Reply:
There are `already` mixed-case labels at the moment (see point 4 in https://github.com/BlueBrain/Search/issues/335#issuecomment-829454811).

We agreed that the `convention` for labels is the `upper` case. Besides, this is the majority use amongst labels.

When the label is in `lower` case, this is then an `issue` to be addressed. Actually, this is done at several places `already` (see below). The goal of clean.py is to meet `best practices` and deal with the case `one and for all` at the source to `avoid` errors and `patches` in the code of BBS. 

<details>
  <summary>grep -r --include "*.py" "upper()" . | sort</summary>

```
./data_and_models/pipelines/ner/add_er.py:    etype_mapping = {external.upper(): params[external]["etype_name"] for external in external_etypes}
./data_and_models/pipelines/ner/eval.py:            )[args.etype.upper()]
./data_and_models/pipelines/ner/eval.py:                etypes_map={args.etype.upper(): params["etype_name"]},
./data_and_models/pipelines/ner/transformers_vs_spacy/transformers/francesco_script.py:                    ent_label = ent["label"].upper()
./src/bluesearch/mining/eval.py:                    ent_label = ent["label"].upper()
```

</details>

Reply:
Otherwise, I don't understand why `if not label.isupper()` would be better than `if label.islower()`.

Indeed, there is `only two` possibilities (upper or lower). So, the two are `logically equivalent`.

Or is there some implementation details giving another conclusion?

Besides, using `if not` tests is considered a bad practice. Mainly because that hinders readability.

Reply:
I don't think they are equivalent. Mixed-case strings give `False` for both:
```python
>>> text = "Hello"
>>> text.islower()
False
>>> text.isupper()
False
>>> not text.isupper()
True
```

Reply:
Oh. I have read "mixed-case labels" as _different_ labels with different cases.

Fair point then! I am adding it to the changes to be done.

Reply:
Done in 29795f22c5d2024f1d81ecc2fb312a7766a8dda6.

Resolving the discussion to make reading the reviews easier. Feel free to reopen if needed.